### ISSUE TITLE
Allow passing dataclass kwargs to the underlying dataclass

### ISSUE THREAD

**Issue by Corentin-Br** [strawberry-graphql/strawberry#2688]
## Feature Request Type

- [ ] Core functionality
- [x] Alteration (enhancement/optimization) of existing feature(s)
- [ ] New behavior

## Description

Currently `_wrap_dataclass` doesn't allow setting kwargs to the dataclass it creates. This leads to the creation (among other things) of a `__repr__` method, which may not be desired.

Case in point, I have classes that are SQLAlchemy models and wrapped with `type` and who inherits a generic `__repr__`.
However, the underlying dataclass creates a new `__repr__` method which makes me load all my attributes. (Which is obviously undesired).

**Comment by ThirVondukr** [comment]
Are you using sqlalchemy dataclasses integration and wrapping it in a `strawberry.type`? I'd still have two separate models here 🤔

**Comment by Corentin-Br** [comment]
We're still on SQLAlchemy 1.4 for now.

Having two models means having to maintain both and needing a way to go from one to the other.
It's also easy to modify a SQLAlchemy model and forget to update the schema model in turn.

We worked with two models for a while and for now we believe having a single unified model will be easier. We'll see down the line.

**Comment by ThirVondukr** [comment]
Your database model and graphql schema are different entities and shouldn't directly depend on each other 🤔 
It's easy for an opposite situation to occur: You can add a private field to your class and it would be automatically exposed in your schema.
You also lose a lot of flexibility in how you can map your db models to strawberry types.

I usually do something like this:
```py
class DbModel(Base):
    ...
    
@strawerry.type
class StrawberryType:
    id: strawberry.ID
    name: str
    price: int
    
    @classmethod
    def from_orm(cls, model: DbModel) -> Self:
        return cls(
            id=strawberry.ID(str(model.id)),
            name=model.name,
            price=model.price,
        )
```
This way you can map your models however you want, and your graphql models are explicitly defined.
To summarize: I think having one entity handle multiple concerns like database schema and your api's schema is a bad idea.

**Comment by ThirVondukr** [comment]
Here's an example, but it's a bit outdated: https://gitlab.com/manga-microservices/manga/-/blob/main/src/gql/modules/manga/types.py

**Comment by Corentin-Br** [comment]
I appreciate your concern, and I do understand those issues, but from our point of view your way is just too much boilerplate to maintain.

Regardless of our specific usecase, I believe being able to pass kwargs is a good thing :)

**Comment by erikwrede** [comment]
Is strawberry-sqlalchemy an option for you? 

Regardless of that, it seems like we dropped the ball on jkimbo's PR a bit. Seems like a feature that has use cases and if we manage to fix the issues in the PR, it could be worth implementing 😊

**Comment by Corentin-Bravo** [comment]
No, strawberry-sqlalchemy is not an option. We have our own strawberry/SQLAlchemy adapter (But I need to look into strawberry-sqlalchemy just for knowledge).

Furthermore even if strawberry-sqlalchemy was an option, the issue is broader than just SQLAlchemy. :)

**Comment by patrick91** [comment]
@Corentin-Bravo what arguments would you like to pass? 

I'm asking because this is easy to do, but at the same time I wonder if we ever switched to something else we'd be doing a bigger breaking change that we'd expect

**Comment by Corentin-Bravo** [comment]
What I need is `repr`, `eq` and `kw_only`
But honestly, I think we should be able to pass all dataclass kwargs (unless there were some that would be specifically breaking with strawberry)

My current "workaround" is to not use the `type` decorator on my dataclasses, but just use the `_process_type` function.

If you're worried that allowing to pass dataclass kwargs could lead to a breaking change in the future, it could also be possible to:
- make _process_type part of the public API.
OR
- Add a kwargs in `type` to allow to bypass all the dataclass wrapping.

I'm not familiar enough with the internals of strawberry to be able to say which way is the best.

**Comment by patrick91** [comment]
> What I need is `repr`, `eq` and `kw_only` But honestly, I think we should be able to pass all dataclass kwargs (unless there were some that would be specifically breaking with strawberry)

Why? 👀 I think we also default to kw_only to be True (are you setting it to false?)


> My current "workaround" is to not use the `type` decorator on my dataclasses, but just use the `_process_type` function.
> 
> If you're worried that allowing to pass dataclass kwargs could lead to a breaking change in the future, it could also be possible to:
> 
> * make _process_type part of the public API.

Mmh this could be an option 🤔 if we changed `_process_type` to something like `process_dataclass` (thought it would create a dataclass, so the name is not great, but you get the idea, to have a dataclass specific processor)

>   OR
> * Add a kwargs in `type` to allow to bypass all the dataclass wrapping.
> 
> I'm not familiar enough with the internals of strawberry to be able to say which way is the best.

Adding kwargs means we need to support them forever pretty much, that's why I'm trying to understand the use cases (since I don't think this has been requested many times)
